### PR TITLE
improve GUI consistency & UX for data-defined style

### PR DESCRIPTION
**Pull request by vmora**
fixes issue 9881 

This PR replaces #1833

### CODE REVIEW

**Comment by nyalldawson** [comment]
@vmora - thanks, will review ASAP

File: src/gui/symbology-ng/qgssymbollayerv2widget.h [hunk 1]
```diff
@@ -17,10 +17,15 @@
 #ifndef QGSSYMBOLLAYERV2WIDGET_H
 #define QGSSYMBOLLAYERV2WIDGET_H
 
+#include <qgsdatadefinedbutton.h>
+
 #include <QWidget>
+#include <QMap>
+#include <QMap>
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbollayerv2widget.h]: duplicate line

File: src/gui/symbology-ng/qgssymbollayerv2widget.h [hunk 2]
```diff
@@ -17,10 +17,15 @@
...
 
 class QgsSymbolLayerV2;
 class QgsVectorLayer;
+class QgsDataDefinedButton;
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbollayerv2widget.h]: shouldn't be required given that the whole header is included

File: src/gui/symbology-ng/qgssymbollayerv2widget.cpp [hunk 3]
```diff
@@ -45,6 +45,28 @@
...
+{
+  symbolLayer()->removeDataDefinedProperties();
+  for ( QMap< QString, QgsDataDefinedButton* >::const_iterator it
+        = mDataDefinedPropertyButtons.begin(); it != mDataDefinedPropertyButtons.end(); ++it )
+  {
+    if ( it.value()->isActive() )
+      symbolLayer()->setDataDefinedProperty( it.key(), it.value()->currentDefinition() );
+  }
+  emit changed();
+}
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbollayerv2widget.cpp]: @vmora Here's a simpler approach: 
```C++
void QgsSymbolLayerV2Widget::registerDataDefinedButton( QgsDataDefinedButton * button, const QString & propertyName, QgsDataDefinedButton::DataType type, const QString & description )
{
  QgsDataDefined dd( symbolLayer()->dataDefinedProperty( propertyName ) );
  button->init( mVectorLayer, &dd, type, Qt::escape( description ) );
  button->setProperty( "propertyName", propertyName );
  connect( button, SIGNAL( dataDefinedChanged( const QString& ) ), this, SLOT( updateDataDefinedProperty() ) );
  connect( button, SIGNAL( dataDefinedActivated( bool ) ), this, SLOT( updateDataDefinedProperty() ) );
}

void QgsSymbolLayerV2Widget::updateDataDefinedProperty()
{
  QgsDataDefinedButton* button = qobject_cast<QgsDataDefinedButton*>( sender() );
  QString propertyName = button->property( "propertyName" ).toString();
  if ( button->isActive() )
  {
    symbolLayer()->setDataDefinedProperty( propertyName, button->currentDefinition() );
  }
  else
  {
    symbolLayer()->removeDataDefinedProperty( propertyName );
  }
  emit changed();
}
```

This avoids the need to remove all properties and rebuild them, and means that mDataDefinedPropertyButtons is no longer required.

**Comment by nyalldawson** [comment]
@vmora Looks much better, thanks. See my note about removing mDataDefinedPropertyButtons - otherwise this is good to merge. A better (more descriptive) commit log message would also be appreciated. Great work!

**Comment by vmora** [comment]
@nyalldawson thanks for the suggestion, it really improves the design.

I triple checked my registered data-defined buttons and noticed that the marker outline style for ellipse was not registered. This is because the expression was not set in the code before, so I fixed that.

**Comment by NathanW2** [comment]
How close to merge is this?

**Comment by vmora** [comment]
qgis_ziplayertest fails on Travis, but seems unrelated.

**Comment by nyalldawson** [comment]
Manually squashed and merged - thank you!

**Pull request closed without merging**